jenkinshelper.ksh

```diff
@@ -0,0 +1,139 @@
+#!/bin/ksh
+# 2021-04-07 Olaf Bohlen <olbohlen@eenfach.de>
+# instead of putting all this into the Jenkinsfile I decided to put the actual code in this script
+
+# global config
+HTTPCONF="/etc/apache2/2.4/conf.d/pkgdepotd.conf"
+
+# just run prepare once to initially set up the environment
+prepare() {
+    echo "jenkinshelper: preparing..."
+    pkg install web/server/apache-24
+    mkdir -p /etc/apache2/2.4/conf.d && chown root:sys /etc/apache2/2.4/conf.d
+    grep "#ProxyPassMatch /default/(.*)\$ http://127.0.0.1:10000/\$1 nocanon max=200" "${HTTPCONF}" >/dev/null 2>&1
+    if [ $? -gt 0 ]; then
+	echo "jenkinshelper: Initializing a new apache config at ${HTTPCONF}"
+	echo "#ProxyPassMatch /default/(.*)\$ http://127.0.0.1:10000/\$1 nocanon max=200" >"${HTTPCONF}"
+    else
+	echo "jenkinshelper: Preserving an existing ${HTTPCONF}"
+    fi
+
+    cat >/etc/apache2/2.4/conf.d/00-proxy.conf <<EOF
+LoadModule proxy_module libexec/mod_proxy.so
+LoadModule proxy_connect_module libexec/mod_proxy_connect.so
+LoadModule proxy_ftp_module libexec/mod_proxy_ftp.so
+LoadModule proxy_http_module libexec/mod_proxy_http.so
+LoadModule proxy_ajp_module libexec/mod_proxy_ajp.so
+
+ProxyRequests Off
+ProxyVia Off
+ProxyPreserveHost on
+RequestHeader unset Origin
+AllowEncodedSlashes NoDecode
+EOF
+    
+    chown root:builders "${HTTPCONF}"
+    chmod 664 "${HTTPCONF}"
+    svcadm enable svc:/network/http:apache24
+}
+
+# setup oi-userland
+stage_setup() {
+    if [ ! -f /etc/apache2/2.4/conf.d/00-proxy.conf ]; then
+	echo "jenkinshelper: aborting, please run \"jenkinshelper -p\" initially as root on this jenkins instance once"
+	exit 1
+    fi
+    echo "jenkinshelper: running gmake setup"
+    gmake setup
+}
+
+# scan the git log for changed components
+# we try to be smart and assume that all updates will always touch
+# the components Makefile also (to update COMPONENT_REVISION, etc)
+stage_build_changed() { 
+    for f in $(git diff --name-only HEAD..origin/oi/hipster | grep Makefile); do
+	echo "jenkinshelper: building for ${f%/*}..."
+	curpwd=$(pwd)
+	cd "${f%/*}" && gmake publish
+        rc=$?
+	cd "${curpwd}"
+	echo "jenkinshelper: done with ${f%/*} return code ${rc}"
+    done
+}
+
+# prepare the pkg.depotd server instance, if an instance already
+# exists for this branch, we will use that - otherwise create a
+# fresh one
+stage_prepare_pkgdepotd()
+{
+    # we need the platform for the path to the repo
+    platform=$(uname -p)
+
+    # check if we already have this branch set up as a pkg.depotd:
+    if ! svcs pkg/server:${BRANCH_NAME} >/dev/null 2>&1; then
```

Comment:
Oh, the horrors: I did find an un-quoted variable expansion! (Maybe ksh it safer in this regard, what it considers a token after expansion, if the var value had a space, but better safe than sorry) :)

Reply:
Thanks